compactor/src/handler.rs

```diff
@@ -414,6 +406,167 @@ async fn compact_hot_partitions(compactor: Arc<Compactor>) -> usize {
     n_candidates
 }
 
+// For a given list of hot partition candidates and a memory budget, compute memory needed to compact each one
+// and compact as many of them in parallel as possible until all candidates are compacted
+async fn compact_hot_partition_candidates(
+    compactor: Arc<Compactor>,
+    mut candidates: Vec<PartitionCompactionCandidateWithInfo>,
+    table_columns: HashMap<TableId, Vec<ColumnTypeCount>>,
+) {
+    let mut remaining_budget_bytes = compactor.config.memory_budget_bytes;
+    let mut parallel_compacting_candidates = vec![];
+
+    while !candidates.is_empty() {
+        // Algorithm:
+        // 1. Remove the first candidate from the list
+        // 2. Check if the candidate can be compacted fully (all L0s and their overlapped L1s) or
+        //    partially (first L0s and their overlapped L1s) under the remaining budget
+        // 3. If yes, add the candidate and its partially/fully compacting files into the `parallel_compacting_candidates`
+        //    Otherwise:
+        //      . if the remaining budget is not the full budget, push the candidate back into the `candidates`
+        //        to consider compacting with larger budget.
+        //      . otherwise, log the partition that it is too large to even compact 2 files
+        // 4. If hit the budget, compact all canidates in the compacting_list in parallel
+        // 5. Repeat
+
+        // --------------------------------------------------------------------
+        // 1. Remove first candidate from the list
+        let partition = candidates.remove(0);
+        let partition_id = partition.candidate.partition_id;
+        let table_id = partition.candidate.table_id;
+
+        // Get column types and their counts for the table of the partition
+        let columns = table_columns.get(&partition.table_id());
```

Comment:
Can this be

```suggestion
        let columns = table_columns.get(&table_id);
```

?